__author__ = 'AJ'
import requests, json, pprint
HEADERS = {'accept': 'application/json'}
def get_exp_url():
    """
    :param search: The endpoint url for searching encode.
    This script will take in a URL from encode and return a list of experiment URLS
    :return: exp_url_list
    """
    base_url = "https://www.encodeproject.org/"
    #search = "search/?type=experiment&assay_term_name=RNA-seq&assembly=hg19&limit=all"
    #search = "/search/?type=experiment&assembly=hg19&limit=all"
    #search = "/search/?searchTerm=EFO%3A0002784&type=experiment&assay_term_name=RNA-seq&status=released&assembly=hg19"
    search = "/search/?searchTerm=EFO%3A0002784&type=experiment&assay_term_name=ChIP-seq&assembly=hg19&status=released"
    url = base_url+search
    r = requests.get(url, headers=HEADERS)

    encode_dict = r.json()

    exp_list = [] # contains end point
    exp_url_list = [] # contains full URL for experiment

    for elem in encode_dict["@graph"]:
        for k in elem:
            if elem["@id"] not in exp_list:
                exp_list.append(elem["@id"])

    for elem in exp_list:
        exp_url_list.append(base_url + elem)

    return exp_url_list


def metadata_extractor(exp_url_list):
    """
    This function extracts relevant metadata fields from an experiment and constructs a TSV file named
    that contains metadata for each experiment URL from encode
    :param exp_url_list: The Experiment list generated by get_exp_url
    :return: None
    : Write: metadata_data.txt
    """
    # Current Version must be updata
    with open('metadata_encode.txt', 'w') as metadata:
        metadata.write(
            "Name\tExperiment_Name\tDescription\tAssay\tBiological_Replicate\tTechnical_Replicate\t"
            "Cell_Type\tBio_Sample\tTarget\tAssembly\tGenome_Annotation\tOutput_Type\tLab\tDate"
            "\tVersion\tSource\tDownload_Link\tSource_Link\tSequencer\tRun_Type\tFile_Type\tBiosample_term_id"
            "\tFile_Size\n")

        for elem in exp_url_list:
            response = requests.get(elem, headers=HEADERS)
            exp_dict = response.json()

            i = 0

            while i < len(exp_dict["files"]): # Checks to see if annotation is present
                if exp_dict["files"][i]["file_type"] == "bam": #and exp_dict["files"][i]["output_type"] == "alignments":


                    # Initialize data_dict and data_dict keys if the file format is bam and is aligned to HG19 reference
                    data_dict = {}

                    data_dict.fromkeys(
                        ["Filename", "Name", "Description", "Assay", "Biological Replicate", "Technical Replicate",
                         "Cell_type", "Bio_Sample", "Target",
                         "Assembly","Genome_Annotation","output_type", "Lab", "Date", "Version", "Source", "Download_Link",
                         "Source_Link", "Sequencer", "Run_Type", "File_Type", "Biosample_term_id", "File_Size"])

                    # Start Data Extraction
                    ###################################################################################################

                    # Filename
                    data_dict["Filename"] = exp_dict["files"][i]["href"]
                    filename = data_dict["Filename"]
                    temp = filename.split("/")
                    metadata.write(temp[4] + "\t")

                    # Name
                    metadata.write(exp_dict["accession"] + "\t")

                    # Description
                    try:
                        if exp_dict["files"][i]["replicate"]["experiment"]["description"] != "":
                            metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["description"].replace("\n","").replace("\r","") + "\t")
                        else:
                            metadata.write("Not Listed" + "\t")
                    except KeyError:
                        if exp_dict["description"] != "":
                            metadata.write(exp_dict["description"] + "\t")
                        else:
                            metadata.write("Not Listed" + "\t")

                    # Assay
                    try:
                        if exp_dict["files"][i]["replicate"]["experiment"]["assay_term_name"] != "":
                            metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["assay_term_name"] + "\t")
                        else:
                            metadata.write("Not Listed" + "\t")
                    except KeyError:
                        if exp_dict["assay_term_name"] != "":
                            metadata.write(exp_dict["assay_term_name"] + "\t")
                        else:
                            metadata.write("Not Listed" + "\t")

                    # Biological Replicate
                    try:
                        metadata.write(str(exp_dict["files"][i]["replicate"]["biological_replicate_number"]) + "\t")

                    except KeyError:
                        metadata.write("Not Listed" + "\t")

                    # Technical Replicate
                    try:
                        metadata.write(str(exp_dict["files"][i]["replicate"]["technical_replicate_number"]) + "\t")
                    except KeyError:
                        metadata.write("Not Listed" + "\t")

                    # Cell_Type
                    try:
                        metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["biosample_term_name"] + "\t")
                    except KeyError:
                        metadata.write(exp_dict["biosample_term_name"] + "\t")

                    # Bio_Sample
                    try:
                        metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["biosample_type"] + "\t")
                    except KeyError:
                        metadata.write(exp_dict["biosample_type"] + "\t")

                    # Target
                    try:
                        metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["target"]["label"] + "\t")
                    except KeyError:
                        metadata.write("Not Listed" + "\t")

                    # Assembly
                    try:
                        metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["assembly"][0] + "\t")
                    except (KeyError, IndexError):
                        metadata.write("not listed" + "\t")

                    # Genome Annotation
                    try:
                        metadata.write(exp_dict["files"][i]["genome_annotation"] + "\t")
                    except (KeyError, IndexError):
                        metadata.write("not listed" + "\t")

                    #output_type
                    try:
                        metadata.write(exp_dict["files"][i]["output_type"] + "\t")

                    except (KeyError,IndexError):
                        metadata.write("not listed" + "\t")

                    # Lab
                    try:
                        metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["lab"]["title"] + "\t")
                    except KeyError:
                        metadata.write(exp_dict["files"][i]["lab"]["title"] + "\t")

                    # Date
                    try:
                        metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["date_released"] + "\t")
                    except KeyError:
                        metadata.write(exp_dict["files"][i]["date_created"] + "\t")

                    # Version
                    Version = data_dict["Version"] = "1"
                    metadata.write(Version + "\t")

                    # Source
                    metadata.write(exp_dict["award"]["project"] + "\t")

                    # Download_Link
                    metadata.write("https://www.encodeproject.org" + exp_dict["files"][i]["href"] + "\t")

                    # Source_Link
                    metadata.write(elem + "\t")

                    # Sequencer
                    try:
                        metadata.write("Not Listed" + "\t")
                    except KeyError:
                        metadata.write(exp_dict["files"][i]["platform"]["title"] + "\t")

                    # Run_Type
                    try:
                        metadata.write(exp_dict["files"][i]["run_type"] + "\t")
                    except KeyError:
                        metadata.write("not listed" + "\t")

                    # File_Type
                    metadata.write(exp_dict["files"][i]["file_format"] + "\t")

                    # Biosample_Term_ID
                    try:
                        metadata.write(exp_dict["files"][i]["replicate"]["experiment"]["biosample_term_id"] + "\t")
                    except KeyError:
                        metadata.write(exp_dict["biosample_term_id"] + "\t")


                    # File_Size
                    fsize = (exp_dict["files"][i]["file_size"]*9.5367431640625e-07)
                    # metadata.write(str(fsize) + "\t")
                    metadata.write(str(fsize))

                    # End of Metadata File Entry
                    metadata.write("\n")
                    metadata.close() #may break code and needs to be testd
                i = i + 1

    return


# Sudo Main
'''
Order of operations to create a tsv file that contains the metadata from encode

Final Output is a new directory named json with json files that represent indivual experiments.
'''

exp_url_list = get_exp_url()
metadata_extractor(exp_url_list)








